components/motor/gpio/basic_test.go

```diff
@@ -318,6 +319,68 @@ func TestMotorABNoEncoder(t *testing.T) {
 	})
 }
 
+func TestGoForInterruption(t *testing.T) {
+	b := &fakeboard.Board{GPIOPins: map[string]*fakeboard.GPIOPin{}}
+	logger := golog.NewTestLogger(t)
+
+	mc := config.Component{
+		Name: "abc",
+	}
+
+	m, err := NewMotor(b, Config{
+		Pins:   PinConfig{A: "1", B: "2", PWM: "3"},
+		MaxRPM: maxRPM, PWMFreq: 4000,
+	}, mc.Name, logger)
+	test.That(t, err, test.ShouldBeNil)
+
+	_, waitDur := goForMath(maxRPM, -50, 100)
+	errChan := make(chan error)
+	startTime := time.Now()
+	go func() {
+		goForErr := m.GoFor(context.Background(), -50, 100, nil)
+		errChan <- goForErr
+	}()
+	testutils.WaitForAssertion(t, func(tb testing.TB) {
+		tb.Helper()
+		isOn, _, onErr := m.IsPowered(context.Background(), nil)
+		test.That(tb, isOn, test.ShouldBeTrue)
+		test.That(tb, onErr, test.ShouldBeNil)
+	})
+	test.That(t, m.Stop(context.Background(), nil), test.ShouldBeNil)
+	receivedErr := <-errChan
+	test.That(t, receivedErr, test.ShouldBeNil)
+	test.That(t, time.Since(startTime), test.ShouldBeLessThan, waitDur)
+	test.That(t, mustGetGPIOPinByName(b, "1").Get(context.Background()), test.ShouldEqual, false)
+	test.That(t, mustGetGPIOPinByName(b, "2").Get(context.Background()), test.ShouldEqual, false)
+	test.That(t, mustGetGPIOPinByName(b, "3").PWM(context.Background()), test.ShouldEqual, 0)
+
+	m, err = NewMotor(b, Config{
+		Pins:   PinConfig{Direction: "1", EnablePinLow: "2", PWM: "3"},
+		MaxRPM: maxRPM, PWMFreq: 4000,
+	}, mc.Name, logger)
+	test.That(t, err, test.ShouldBeNil)
+
+	_, waitDur = goForMath(maxRPM, 50, 100)
+	errChan = make(chan error)
+	startTime = time.Now()
+	go func() {
+		goForErr := m.GoFor(context.Background(), 50, 100, nil)
+		errChan <- goForErr
+	}()
+	testutils.WaitForAssertion(t, func(tb testing.TB) {
+		tb.Helper()
+		isOn, _, onErr := m.IsPowered(context.Background(), nil)
+		test.That(tb, onErr, test.ShouldBeNil)
+		test.That(tb, isOn, test.ShouldBeTrue)
+	})
+	test.That(t, m.Stop(context.Background(), nil), test.ShouldBeNil)
+	receivedErr = <-errChan
+	test.That(t, receivedErr, test.ShouldBeNil)
+	test.That(t, time.Since(startTime), test.ShouldBeLessThan, waitDur)
+	test.That(t, mustGetGPIOPinByName(b, "1").Get(context.Background()), test.ShouldEqual, true)
+	test.That(t, mustGetGPIOPinByName(b, "3").PWM(context.Background()), test.ShouldEqual, 0)
```

Comment:
These two paragraphs seem identical to the previous two, except we don't test pin 2 and the we GoFor 50 rpm instead of -50. Does Go's testing framework have a way to abstract out the shared code so it's clearer what's going on? If this were Python, I'd say to use the `parameterized` module, but Go doesn't obviously have something analogous.

I'd prefer to have DRYer code, so it's clearer what is changing from one section to the next (easier to read/understand), and you can add extra parts to the test more easily (for example, I wonder if you started by never testing pin 2, copied and pasted the code for the second half, and then added in pin 2 to the first half but forgot to do it here. Was that an oversight or on purpose?). but I acknowledge that I don't see a popular way to avoid copying and pasting like this, and maybe the Go community thinks that's the right approach.

Reply:
there's a way, I can update

Reply:
I take that back actually, the test for expected pin states being so different makes it annoying enough to not be worth it. I actually tend to not DRY up test code as much anyway because it makes it easier to track the line number of a failing case